apps/sms/services/test/unit/moz_mobile_message/moz_mobile_message_shim_test.js

```diff
@@ -0,0 +1,375 @@
+  });
+
+  test('method mapping is correctly initialized', function() {
+    METHODS.forEach((shimMethod) => {
+      sinon.assert.calledWith(
+        serviceStub.method,
+        shimMethod,
+        MozMobileMessageShim[shimMethod]
+      );
+    });
+  });
+
+  test('stream mapping is correctly initialized', function() {
+    STREAMS.forEach((shimStream) => {
+      sinon.assert.calledWith(
+        serviceStub.stream,
+        shimStream,
+        MozMobileMessageShim[shimStream]
+      );
+    });
+  });
+
+  suite('event broadcast', function() {
+    var onEvent;
+
+    setup(function() {
+      onEvent = {
+        message: {},
+        deletedThreadIds: [1, 2, 3]
+      };
+    });
+
+    test('onSending', function() {
+      MozMobileMessageShim.onSending(onEvent);
+
+      sinon.assert.calledWith(
+        serviceStub.broadcast,
+        'message-sending',
+        { message: onEvent.message }
+      );
+    });
+
+    test('onFailed', function() {
+      MozMobileMessageShim.onFailed(onEvent);
+
+      sinon.assert.calledWith(
+        serviceStub.broadcast,
+        'message-failed-to-send',
+        { message: onEvent.message }
+      );
+    });
+
+    test('onDeliverysuccess', function() {
+      MozMobileMessageShim.onDeliverysuccess(onEvent);
+
+      sinon.assert.calledWith(
+        serviceStub.broadcast,
+        'message-delivered',
+        { message: onEvent.message }
+      );
+    });
+
+    test('onReadsuccess', function() {
+      MozMobileMessageShim.onReadsuccess(onEvent);
+
+      sinon.assert.calledWith(
+        serviceStub.broadcast,
+        'message-read',
+        { message: onEvent.message }
+      );
+    });
+
+    test('onSent', function() {
+      MozMobileMessageShim.onSent(onEvent);
+
+      sinon.assert.calledWith(
+        serviceStub.broadcast,
+        'message-sent',
+        { message: onEvent.message }
+      );
+    });
+
+    test('onReceived', function() {
+      MozMobileMessageShim.onReceived(onEvent);
+
+      sinon.assert.calledWith(
+        serviceStub.broadcast,
+        'message-received',
+        { message: onEvent.message }
+      );
+    });
+
+    test('onDeleted', function() {
+      MozMobileMessageShim.onDeleted(onEvent);
+
+      sinon.assert.calledWith(
+        serviceStub.broadcast,
+        'threads-deleted',
+        { ids: onEvent.deletedThreadIds }
+      );
+    });
+  });
+
+  suite('methods for wrapping up API', function() {
+
+    setup(function() {
+      this.sinon.spy(MockNavigatormozMobileMessage, 'getMessage');
+      this.sinon.spy(MockNavigatormozMobileMessage, 'retrieveMMS');
+      this.sinon.spy(MockNavigatormozMobileMessage, 'send');
+      this.sinon.spy(MockNavigatormozMobileMessage, 'sendMMS');
+      this.sinon.spy(MockNavigatormozMobileMessage, 'markMessageRead');
+      this.sinon.spy(MockNavigatormozMobileMessage, 'getSegmentInfoForText');
+      MockNavigatormozMobileMessage.delete = sinon.stub();
+    });
+
+    test('getMessage', function() {
```

Comment:
I know these test look so many dup part and it should be able to simplify into a loop, but I think we might not able to return directly before bug 1172794 landed, so still keep the test in totally independent format.